File: client/src/game/Scenario.ts
import Aircraft from "@/game/units/Aircraft";
import Airbase from "@/game/units/Airbase";
import Facility from "@/game/units/Facility";
import Side from "@/game/Side";
import Weapon from "@/game/units/Weapon";
import Ship from "@/game/units/Ship";
import { getDistanceBetweenTwoPoints } from "@/utils/mapFunctions";
import ReferencePoint from "@/game/units/ReferencePoint";
import PatrolMission from "@/game/mission/PatrolMission";
import { Target } from "@/game/engine/weaponEngagement";
import StrikeMission from "@/game/mission/StrikeMission";
import { Mission } from "@/game/Game";
import { SIDE_COLOR } from "@/utils/colors";
import Relationships from "@/game/Relationships";
import { randomUUID } from "@/utils/generateUUID";

type HomeBase = Airbase | Ship;

interface IScenario {
  id: string;
  name: string;
  startTime: number;
  currentTime?: number;
  duration: number;
  sides?: Side[];
  timeCompression?: number;
  aircraft?: Aircraft[];
  ships?: Ship[];
  facilities?: Facility[];
  airbases?: Airbase[];
  weapons?: Weapon[];
  referencePoints?: ReferencePoint[];
  missions?: PatrolMission[];
  relationships?: Relationships;
}

export default class Scenario {
  id: string;
  name: string;
  startTime: number;
  currentTime: number;
  duration: number;
  sides: Side[];
  timeCompression: number;
  aircraft: Aircraft[];
  ships: Ship[];
  facilities: Facility[];
  airbases: Airbase[];
  weapons: Weapon[];
  referencePoints: ReferencePoint[];
  missions: Mission[];
  relationships: Relationships;

  constructor(parameters: IScenario) {
    this.id = parameters.id;
    this.name = parameters.name;
    this.startTime = parameters.startTime;
    this.currentTime = parameters.currentTime ?? parameters.startTime;
    this.duration = parameters.duration;
    this.sides = parameters.sides ?? [];
    this.timeCompression = parameters.timeCompression ?? 1;
    this.aircraft = parameters.aircraft ?? [];
    this.facilities = parameters.facilities ?? [];
    this.airbases = parameters.airbases ?? [];
    this.weapons = parameters.weapons ?? [];
    this.ships = parameters.ships ?? [];
    this.referencePoints = parameters.referencePoints ?? [];
    this.missions = parameters.missions ?? [];
    this.relationships = parameters.relationships ?? new Relationships({});
  }

  getSide(sideId: string | null | undefined): Side | undefined {
    return this.sides.find((side) => side.id === sideId);
  }

  getSideName(sideId: string | null | undefined): string {
    const side = this.getSide(sideId);
    if (side) {
      return side.name;
    }
    return "N/A";
  }

  getSideColor(sideId: string | null | undefined): SIDE_COLOR {
    const side = this.getSide(sideId);
    if (side) {
      return side.color;
    }
    return SIDE_COLOR.BLACK;
  }

  getAircraft(aircraftId: string | null): Aircraft | undefined {
    return this.aircraft.find((aircraft) => aircraft.id === aircraftId);
  }

  getFacility(facilityId: string | null): Facility | undefined {
    return this.facilities.find((facility) => facility.id === facilityId);
  }

  getAirbase(airbaseId: string | null): Airbase | undefined {
    return this.airbases.find((airbase) => airbase.id === airbaseId);
  }

  getWeapon(weaponId: string | null): Weapon | undefined {
    return this.weapons.find((weapon) => weapon.id === weaponId);
  }

  getShip(shipId: string | null): Ship | undefined {
    return this.ships.find((ship) => ship.id === shipId);
  }

  getReferencePoint(
    referencePointId: string | null
  ): ReferencePoint | undefined {
    return this.referencePoints.find(
      (referencePoint) => referencePoint.id === referencePointId
    );
  }

  getPatrolMission(missionId: string | null): PatrolMission | undefined {
    return this.missions.find(
      (mission) => mission.id === missionId && mission instanceof PatrolMission
    ) as PatrolMission;
  }

  getStrikeMission(missionId: string | null): StrikeMission | undefined {
    return this.missions.find(
      (mission) => mission.id === missionId && mission instanceof StrikeMission
    ) as StrikeMission;
  }

  getAllPatrolMissions(): PatrolMission[] {
    return this.missions.filter(
      (mission) => mission instanceof PatrolMission
    ) as PatrolMission[];
  }

  getAllStrikeMissions(): StrikeMission[] {
    return this.missions.filter(
      (mission) => mission instanceof StrikeMission
    ) as StrikeMission[];
  }

  getMissionByAssignedUnitId(unitId: string): Mission | undefined {
    return this.missions.find((mission) =>
      mission.assignedUnitIds.includes(unitId)
    );
  }

  updateScenarioName(name: string): void {
    this.name = name;
  }

  deleteWeaponFromAircraft(aircraftId: string, weaponId: string): Weapon[] {
    const aircraft = this.getAircraft(aircraftId);
    let aircraftWeapons: Weapon[] = [];
    if (aircraft) {
      const weaponIndex = aircraft.weapons.findIndex(
        (weapon) => weapon.id === weaponId
      );
      if (weaponIndex !== -1) {
        aircraft.weapons.splice(weaponIndex, 1);
      }
      aircraftWeapons = aircraft.weapons;
    }
    return aircraftWeapons;
  }

  updateAircraftWeaponQuantity(
    aircraftId: string,
    weaponId: string,
    increment: number
  ) {
    const aircraft = this.getAircraft(aircraftId);
    let aircraftWeapons: Weapon[] = [];
    if (aircraft) {
      const weapon = aircraft.weapons.find((weapon) => weapon.id === weaponId);
      if (weapon) {
        weapon.currentQuantity += increment;
        if (weapon.currentQuantity < 0) {
          weapon.currentQuantity = 0;
        }
      }
      aircraftWeapons = aircraft.weapons;
    }
    return aircraftWeapons;
  }

  addWeaponToAircraft(
    aircraftId: string,
    weaponClassName?: string,
    weaponSpeed?: number,
    weaponMaxFuel?: number,
    weaponFuelRate?: number,
    weaponLethality?: number
  ): Weapon[] {
    const aircraft = this.getAircraft(aircraftId);
    let aircraftWeapons: Weapon[] = [];
    if (aircraft) {
      aircraftWeapons = aircraft.weapons;
      if (
        !(
          weaponClassName &&
          weaponSpeed &&
          weaponMaxFuel &&
          weaponFuelRate &&
          weaponLethality
        )
      ) {
        return aircraftWeapons;
      }
      if (
        aircraft.weapons.find((weapon) => weapon.className === weaponClassName)
      ) {
        return aircraftWeapons;
      }
      const weapon = new Weapon({
        id: randomUUID(),
        name: weaponClassName,
        sideId: aircraft.sideId,
        className: weaponClassName,
        latitude: 0.0,
        longitude: 0.0,
        altitude: 10000.0,
        heading: 90.0,
        speed: weaponSpeed,
        currentFuel: weaponMaxFuel,
        maxFuel: weaponMaxFuel,
        fuelRate: weaponFuelRate,
        range: 100,
        sideColor: aircraft.sideColor,
        targetId: null,
        lethality: weaponLethality,
        maxQuantity: 1,
        currentQuantity: 1,
      });
      aircraftWeapons.push(weapon);
    }
    return aircraftWeapons;
  }

  deleteWeaponFromFacility(facilityId: string, weaponId: string): Weapon[] {
    const facility = this.getFacility(facilityId);
    let facilityWeapons: Weapon[] = [];
    if (facility) {
      const weaponIndex = facility.weapons.findIndex(
        (weapon) => weapon.id === weaponId
      );
      if (weaponIndex !== -1) {
        facility.weapons.splice(weaponIndex, 1);
      }
      facilityWeapons = facility.weapons;
    }
    return facilityWeapons;
  }

  updateFacilityWeaponQuantity(
    facilityId: string,
    weaponId: string,
    increment: number
  ) {
    const facility = this.getFacility(facilityId);
    let facilityWeapons: Weapon[] = [];
    if (facility) {
      const weapon = facility.weapons.find((weapon) => weapon.id === weaponId);
      if (weapon) {
        weapon.currentQuantity += increment;
        if (weapon.currentQuantity < 0) {
          weapon.currentQuantity = 0;
        }
      }
      facilityWeapons = facility.weapons;
    }
    return facilityWeapons;
  }

  addWeaponToFacility(
    facilityId: string,
    weaponClassName?: string,
    weaponSpeed?: number,
    weaponMaxFuel?: number,
    weaponFuelRate?: number,
    weaponLethality?: number
  ): Weapon[] {
    const facility = this.getFacility(facilityId);
    let facilityWeapons: Weapon[] = [];
    if (facility) {
      facilityWeapons = facility.weapons;
      if (
        !(
          weaponClassName &&
          weaponSpeed &&
          weaponMaxFuel &&
          weaponFuelRate &&
          weaponLethality
        )
      ) {
        return facilityWeapons;
      }
      if (
        facility.weapons.find((weapon) => weapon.className === weaponClassName)
      ) {
        return facilityWeapons;
      }
      const weapon = new Weapon({
        id: randomUUID(),
        name: weaponClassName,
        sideId: facility.sideId,
        className: weaponClassName,
        latitude: 0.0,
        longitude: 0.0,
        altitude: 10000.0,
        heading: 90.0,
        speed: weaponSpeed,
        currentFuel: weaponMaxFuel,
        maxFuel: weaponMaxFuel,
        fuelRate: weaponFuelRate,
        range: 100,
        sideColor: facility.sideColor,
        targetId: null,
        lethality: weaponLethality,
        maxQuantity: 1,
        currentQuantity: 1,
      });
      facilityWeapons.push(weapon);
    }
    return facilityWeapons;
  }

  updateAircraft(
    aircraftId: string,
    aircraftName: string,
    aircraftClassName: string,
    aircraftSpeed: number,
    aircraftCurrentFuel: number,
    aircraftFuelRate: number
  ) {
    const aircraft = this.getAircraft(aircraftId);
    if (aircraft) {
      aircraft.name = aircraftName;
      aircraft.className = aircraftClassName;
      aircraft.speed = aircraftSpeed;
      aircraft.currentFuel = aircraftCurrentFuel;
      aircraft.fuelRate = aircraftFuelRate;
    }
  }

  updateFacility(
    facilityId: string,
    facilityName: string,
    facilityClassName: string,
    facilityRange: number
  ) {
    const facility = this.getFacility(facilityId);
    if (facility) {
      facility.name = facilityName;
      facility.className = facilityClassName;
      facility.range = facilityRange;
    }
  }

  updateAirbase(airbaseId: string, airbaseName: string) {
    const airbase = this.getAirbase(airbaseId);
    if (airbase) {
      airbase.name = airbaseName;
    }
  }

  updateShip(
    shipId: string,
    shipName: string,
    shipClassName: string,
    shipSpeed: number,
    shipCurrentFuel: number,
    shipWeaponQuantity: number,
    shipRange: number
  ) {
    const ship = this.getShip(shipId);
    if (ship) {
      ship.name = shipName;
      ship.className = shipClassName;
      ship.speed = shipSpeed;
      ship.currentFuel = shipCurrentFuel;
      ship.range = shipRange;
      ship.weapons.forEach((weapon) => {
        weapon.currentQuantity = shipWeaponQuantity;
      });
    }
  }

  updateReferencePoint(referencePointId: string, referencePointName: string) {
    const referencePoint = this.getReferencePoint(referencePointId);
    if (referencePoint) {
      referencePoint.name = referencePointName;
    }
  }

  getAircraftHomeBase(aircraftId: string): HomeBase | undefined {
    const aircraft = this.getAircraft(aircraftId);
    if (aircraft) {
      return (
        this.getAirbase(aircraft.homeBaseId) ??
        this.getShip(aircraft.homeBaseId)
      );
    }
  }

  getClosestBaseToAircraft(aircraftId: string): HomeBase | undefined {
    const aircraft = this.getAircraft(aircraftId);
    if (aircraft) {
      let closestBase: HomeBase | undefined;
      let closestDistance = Number.MAX_VALUE;
      this.airbases.forEach((airbase) => {
        if (airbase.sideId !== aircraft.sideId) return;
        const distance = getDistanceBetweenTwoPoints(
          aircraft.latitude,
          aircraft.longitude,
          airbase.latitude,
          airbase.longitude
        );
        if (distance < closestDistance) {
          closestDistance = distance;
          closestBase = airbase;
        }
      });
      this.ships.forEach((ship) => {
        if (ship.sideId !== aircraft.sideId) return;
        const distance = getDistanceBetweenTwoPoints(
          aircraft.latitude,
          aircraft.longitude,
          ship.latitude,
          ship.longitude
        );
        if (distance < closestDistance) {
          closestDistance = distance;
          closestBase = ship;
        }
      });
      return closestBase;
    }
  }

  getAllTargetsFromEnemySides(sideId: string): Target[] {
    const targets: Target[] = [];
    this.aircraft.forEach((aircraft) => {
      if (this.isHostile(aircraft.sideId, sideId)) {
        targets.push(aircraft);
      }
    });
    this.facilities.forEach((facility) => {
      if (this.isHostile(facility.sideId, sideId)) {
        targets.push(facility);
      }
    });
    this.ships.forEach((ship) => {
      if (this.isHostile(ship.sideId, sideId)) {
        targets.push(ship);
      }
    });
    this.airbases.forEach((airbase) => {
      if (this.isHostile(airbase.sideId, sideId)) {
        targets.push(airbase);
      }
    });
    return targets;
  }

  isHostile(sideId: string, targetId: string): boolean {
    return this.relationships.isHostile(sideId, targetId);
  }
}

File: gym/blade/Scenario.py
import json

from blade.units.Aircraft import Aircraft
from blade.units.Ship import Ship
from blade.units.Facility import Facility
from blade.units.Airbase import Airbase
from blade.units.Weapon import Weapon
from blade.units.ReferencePoint import ReferencePoint
from blade.Side import Side
from blade.mission.PatrolMission import PatrolMission
from blade.mission.StrikeMission import StrikeMission
from blade.utils.utils import get_distance_between_two_points
from blade.utils.colors import SIDE_COLOR
from blade.Relationships import Relationships

HomeBase = Airbase | Ship

Target = Aircraft | Facility | Weapon | Airbase | Ship


class Scenario:
    def __init__(
        self,
        id: str = "",
        name: str = "",
        start_time: int = 0,
        duration: int = 1,
        sides: list[Side] = [],
        current_time: int = None,
        time_compression: int = 1,
        aircraft: list[Aircraft] = None,
        ships: list[Ship] = None,
        facilities: list[Facility] = None,
        airbases: list[Airbase] = None,
        weapons: list[Weapon] = None,
        reference_points: list[ReferencePoint] = None,
        missions: list[PatrolMission | StrikeMission] = None,
        relationships: Relationships = Relationships(),
    ):
        self.id = id
        self.name = name
        self.start_time = start_time
        self.current_time = current_time if current_time is not None else start_time
        self.duration = duration
        self.sides = sides
        self.time_compression = time_compression
        self.aircraft = aircraft if aircraft is not None else []
        self.ships = ships if ships is not None else []
        self.facilities = facilities if facilities is not None else []
        self.airbases = airbases if airbases is not None else []
        self.weapons = weapons if weapons is not None else []
        self.reference_points = reference_points if reference_points is not None else []
        self.missions = missions if missions is not None else []
        self.relationships = relationships

    def get_side(self, side_id: str | None) -> Side | None:
        for side in self.sides:
            if side.id == side_id:
                return side
        return None

    def get_side_name(self, side_id: str | None) -> str:
        side = self.get_side(side_id)
        return side.name if side is not None else "N/A"

    def get_side_color(self, side_id: str | None) -> SIDE_COLOR:
        side = self.get_side(side_id)
        return side.color if side is not None else SIDE_COLOR.BLACK

    def get_aircraft(self, aircraft_id: str) -> Aircraft | None:
        for aircraft in self.aircraft:
            if aircraft.id == aircraft_id:
                return aircraft
        return None

    def get_facility(self, facility_id: str) -> Facility | None:
        for facility in self.facilities:
            if facility.id == facility_id:
                return facility
        return None

    def get_airbase(self, airbase_id: str) -> Airbase | None:
        for airbase in self.airbases:
            if airbase.id == airbase_id:
                return airbase
        return None

    def get_ship(self, ship_id: str) -> Ship | None:
        for ship in self.ships:
            if ship.id == ship_id:
                return ship
        return None

    def get_weapon(self, weapon_id: str) -> Weapon | None:
        for weapon in self.weapons:
            if weapon.id == weapon_id:
                return weapon
        return None

    def get_target(self, target_id: str) -> Target | None:
        for target in (
            self.aircraft + self.ships + self.facilities + self.airbases + self.weapons
        ):
            if target.id == target_id:
                return target
        return None

    def get_reference_point(self, reference_point_id: str) -> ReferencePoint | None:
        for reference_point in self.reference_points:
            if reference_point.id == reference_point_id:
                return reference_point
        return None

    def get_patrol_mission(self, mission_id: str) -> PatrolMission | None:
        for mission in self.missions:
            if isinstance(mission, PatrolMission) and mission.id == mission_id:
                return mission
        return None

    def get_strike_mission(self, mission_id: str) -> StrikeMission | None:
        for mission in self.missions:
            if isinstance(mission, StrikeMission) and mission.id == mission_id:
                return mission
        return None

    def get_all_patrol_missions(self) -> list[PatrolMission]:
        return [
            mission for mission in self.missions if isinstance(mission, PatrolMission)
        ]

    def get_all_strike_missions(self) -> list[StrikeMission]:
        return [
            mission for mission in self.missions if isinstance(mission, StrikeMission)
        ]

    def update_aircraft(
        self,
        aircraft_id: str,
        aircraft_name: str,
        aircraft_class_name: str,
        aircraft_speed: float,
        aircraft_weapon_quantity: float,
        aircraft_current_fuel: float,
        aircraft_fuel_rate: float,
        sample_weapon: Weapon,
    ):
        aircraft = self.get_aircraft(aircraft_id)
        if aircraft is not None:
            aircraft.name = aircraft_name
            aircraft.class_name = aircraft_class_name
            aircraft.speed = aircraft_speed
            if len(aircraft.weapons) < 1:
                aircraft.weapons = [sample_weapon]
            else:
                for weapon in aircraft.weapons:
                    weapon.current_quantity = aircraft_weapon_quantity
            aircraft.current_fuel = aircraft_current_fuel
            aircraft.fuel_rate = aircraft_fuel_rate

    def update_facility(
        self,
        facility_id: str,
        facility_name: str,
        facility_class_name: str,
        facility_range: float,
        facility_weapon_quantity: float,
        sample_weapon: Weapon,
    ):
        facility = self.get_facility(facility_id)
        if facility is not None:
            facility.name = facility_name
            facility.class_name = facility_class_name
            facility.range = facility_range
            if len(facility.weapons) < 1:
                facility.weapons = [sample_weapon]
            else:
                for weapon in facility.weapons:
                    weapon.current_quantity = facility_weapon_quantity

    def update_airbase(
        self,
        airbase_id: str,
        airbase_name: str,
    ):
        airbase = self.get_airbase(airbase_id)
        if airbase is not None:
            airbase.name = airbase_name

    def update_ship(
        self,
        ship_id: str,
        ship_name: str,
        ship_class_name: str,
        ship_speed: float,
        ship_weapon_quantity: float,
        ship_current_fuel: float,
        ship_fuel_rate: float,
        ship_range: float,
        sample_weapon: Weapon,
    ):
        ship = self.get_ship(ship_id)
        if ship is not None:
            ship.name = ship_name
            ship.class_name = ship_class_name
            ship.speed = ship_speed
            if len(ship.weapons) < 1:
                ship.weapons = [sample_weapon]
            else:
                for weapon in ship.weapons:
                    weapon.current_quantity = ship_weapon_quantity
            ship.current_fuel = ship_current_fuel
            ship.fuel_rate = ship_fuel_rate
            ship.range = ship_range

    def update_reference_point(
        self,
        reference_point_id: str,
        reference_point_name: str,
    ):
        reference_point = self.get_reference_point(reference_point_id)
        if reference_point is not None:
            reference_point.name = reference_point_name

    def get_aircraft_homebase(self, aircraft_id: str) -> HomeBase | None:
        aircraft = self.get_aircraft(aircraft_id)
        if aircraft is not None:
            base = self.get_airbase(aircraft.home_base_id)
            if base is not None:
                return base
            else:
                ship = self.get_ship(aircraft.home_base_id)
                if ship is not None:
                    return ship
        return None

    def get_closest_base_to_aircraft(self, aircraft_id: str) -> HomeBase | None:
        aircraft = self.get_aircraft(aircraft_id)
        if aircraft is not None:
            closest_base = None
            closest_distance = float("inf")
            for base in self.airbases + self.ships:
                if base.side_id != aircraft.side_id:
                    continue
                distance = get_distance_between_two_points(
                    aircraft.latitude, aircraft.longitude, base.latitude, base.longitude
                )
                if distance < closest_distance:
                    closest_base = base
                    closest_distance = distance
            return closest_base
        return None

    def get_all_targets_from_enemy_sides(self, side_id: str) -> Target:
        targets = []
        for aircraft in self.aircraft:
            if self.is_hostile(aircraft.side_id, side_id):
                targets.append(aircraft)
        for facility in self.facilities:
            if self.is_hostile(facility.side_id, side_id):
                targets.append(facility)
        for ship in self.ships:
            if self.is_hostile(ship.side_id, side_id):
                targets.append(ship)
        for airbase in self.airbases:
            if self.is_hostile(airbase.side_id, side_id):
                targets.append(airbase)
        return targets

    def is_hostile(self, side_id: str, target_id: str) -> bool:
        return self.relationships.is_hostile(side_id, target_id)

    def to_dict(self):
        def serialize(obj):
            if hasattr(obj, "to_dict"):
                return obj.to_dict()
            elif isinstance(obj, list):
                return [serialize(item) for item in obj]
            elif isinstance(obj, dict):
                return {key: serialize(value) for key, value in obj.items()}
            else:
                return obj

        return serialize(self.__dict__)

    def toJson(self):
        return json.dumps(self.to_dict(), sort_keys=True, indent=4)
